metaflow/runtime.py

```diff
@@ -756,6 +823,67 @@ def log(self, msg, system_msg=False, pid=None, timestamp=True):
         self._logger(msg, head=prefix, system_msg=system_msg, timestamp=timestamp)
         sys.stdout.flush()
 
+    def _get_task_id(self, task_id):
+        already_existed = True
+        if self.ubf_context == UBF_CONTROL:
+            [input_path] = self.input_paths
+            run, input_step, input_task = input_path.split("/")
+            # We associate the control task-id to be 1:1 with the split node
+            # where the unbounded-foreach was defined.
+            # We prefer encoding the corresponding split into the task_id of
+            # the control node; so it has access to this information quite
+            # easily. There is anyway a corresponding int id stored in the
+            # metadata backend - so this should be fine.
+            task_id = "control-%s-%s-%s" % (run, input_step, input_task)
+        # Register only regular Metaflow (non control) tasks.
+        if task_id is None:
+            task_id = str(self.metadata.new_task_id(self.run_id, self.step))
+            already_existed = False
+        else:
+            # task_id is preset only by persist_constants() or control tasks.
+            if self.ubf_context == UBF_CONTROL:
+                tags = [CONTROL_TASK_TAG]
+                attempt_id = 0
+                already_existed = self.metadata.register_task_id(
+                    self.run_id,
+                    self.step,
+                    task_id,
+                    attempt_id,
+                    sys_tags=tags,
+                )
+                # As part of tag mutation project, we will be overlaying Run's tags
+                # (user and system) on top of Task tags.  I.e. a Task will no longer have
+                # its own tags distinct from their ancestral Run.  It means we should not
+                # use tags to indicate if a task is a control task in future.
+                #
+                # This is the main PR that implements the same "stashing to task metadata"
+                # concept:
+                #     https://github.com/Netflix/metaflow/pull/1039
```

Comment:
Given that this is merged and released, we can update the comment.

Reply:
Sure. I had just copy pasted. I'll remove it.